prov/tcp/src/tcpx_progress.c

```diff
@@ -846,3 +813,159 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
 	[ofi_op_read_rsp] = tcpx_op_read_rsp,
 	[ofi_op_write] = tcpx_op_write,
 };
+
+void tcpx_progress_domain(struct tcpx_domain *domain)
+{
+	struct ofi_epollfds_event events[MAX_POLL_EVENTS];
+	struct dlist_entry *item;
+	struct tcpx_ep *ep;
+	struct fid *fid;
+	uint32_t inevent, outevent, errevent;
+	int nfds, i;
+
+	ofi_mutex_lock(&domain->progress_lock);
+	dlist_foreach(&domain->ep_list, item) {
+		ep = container_of(item, struct tcpx_ep, ep_entry);
+
+		ofi_mutex_lock(&ep->lock);
+		/* We need to progress receives in the case where we're waiting
+		 * on the application to post a buffer to consume a receive
+		 * that we've already read from the kernel.  If the message is
+		 * of length 0, there's no additional data to read, so failing
+		 * to progress can result in application hangs.
+		 */
+		if (ofi_bsock_readable(&ep->bsock) ||
+		    (ep->cur_rx.handler && !ep->cur_rx.entry)) {
+			assert(ep->state == TCPX_CONNECTED);
+			tcpx_progress_rx(ep);
+		}
+
+		(void) tcpx_update_epoll(ep);
+		ofi_mutex_unlock(&ep->lock);
+	}
+	ofi_mutex_unlock(&domain->progress_lock);
+
+	if (domain->wait_obj == FI_WAIT_FD) {
+		nfds = ofi_epoll_wait(domain->epoll_fd, events,
+				      MAX_POLL_EVENTS, 0);
+		inevent = OFI_EPOLL_IN;
+		outevent = OFI_EPOLL_OUT;
+		errevent = OFI_EPOLL_ERR;
+	} else {
+		nfds = ofi_pollfds_wait(domain->pollfds, events,
+					MAX_POLL_EVENTS, 0);
+		inevent = POLLIN;
+		outevent = POLLOUT;
+		errevent = POLLERR;
+	}
+	if (nfds <= 0)
+		return;
+
+	for (i = 0; i < nfds; i++) {
+		fid = events[i].data.ptr;
+		if (fid->fclass != FI_CLASS_EP) {
+			fd_signal_reset(&domain->signal);
+			continue;
+		}
+
+		ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid.fid);
+		ofi_mutex_lock(&ep->lock);
+		if (events[i].events & errevent)
+			tcpx_progress_async(ep);
+		if (events[i].events & inevent)
+			tcpx_progress_rx(ep);
+		if (events[i].events & outevent)
+			tcpx_progress_tx(ep);
+		ofi_mutex_unlock(&ep->lock);
+	}
+}
+
+static void *tcpx_auto_progress(void *arg)
+{
+	struct tcpx_domain *domain = arg;
+
+	FI_INFO(&tcpx_prov, FI_LOG_DOMAIN, "progress thread starting\n");
+
+	ofi_mutex_lock(&domain->progress_lock);
+	while (domain->auto_progress) {
+		ofi_mutex_unlock(&domain->progress_lock);
+
+		// TODO: make blocking to avoid spinning
```

Comment:
What strategy do you have in mind to avoid spinning here? I thought about the following code:

```
struct ofi_epollfds_event events[1];

if (domain->wait_obj == FI_WAIT_FD) {
	nfds = ofi_epoll_wait(domain->epoll_fd, events,
			      1, -1);
} else {
	nfds = ofi_pollfds_wait(domain->pollfds, events,
				1, -1);
}

if (nfds > 0)
	tcpx_progress_domain(domain);
```

I think calling `ofi_{epoll/pollfds}_wait()` multiple times without processing the events is ok since the poll set is configured for level-triggered. 
What do you think?

Reply:
I'm hoping something like what you have will work.  I'd like to avoid calling poll() multiple times and discarding the events from one of the calls, since poll() appears to be a slow operation when watching across a large number of sockets.  In that case, we'd need to pass a timeout to the progress_domain() call, but figure out a way to avoid use after close.